### PR TITLE
✨ Add User organization filter

### LINKED ISSUE #700: Filter allUsers by organization
The `allUsers` query should accept an option to filter users for a given organization.

### PR DESCRIPTION
**Pull request by gsantia**
Closes #700 

Add the ability to filter users by organization in the `allUsers` query. An organization node ID is supplied as input to the query and all users that are a member of that organization are returned.

Screen Shot 2021-07-23 at 4 21 04 PM

### CODE REVIEW

File: tests/users/test_queries.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from django.contrib.auth import get_user_model
+from creator.organizations.factories import OrganizationFactory
+from creator.organizations.models import Organization
+from creator.users.factories import UserFactory
+
+User = get_user_model()
+
+ALL_USERS_ORGANIZATION = """
+query ($org_name: String) {
+  allUsers (organizations_Name: $org_name) {
```
Comment (shanewilson) [inline tests/users/test_queries.py:13]: the allStudies query uses an id for organization filters - if might be good to do that here as well:

```
allStudies(organization: ID!) 
```

File: creator/users/schema.py [hunk 2]
```diff
@@ -93,6 +95,12 @@ def _get_study_id(self, info):
 class UserNode(DjangoObjectType):
     display_name = graphene.String(source="display_name")
 
+    organizations = DjangoFilterConnectionField(
```
Comment (shanewilson) [inline creator/users/schema.py:98]: this is good - can we also add `users` sub query to an `organization`?

File: tests/users/test_queries.py (hunk 1 above)
Comment (gsantia) [reply tests/users/test_queries.py:13]: yeah, i wasn't sure about this

File: creator/users/schema.py (hunk 2 above)
Comment (gsantia) [reply creator/users/schema.py:98]: sure, i'll make an issue and work on it and try to get it out today

File: creator/users/schema.py [hunk 3]
```diff
@@ -189,6 +197,7 @@ class Meta:
             "last_name": ["exact"],
             "last_login": ["gt", "lt"],
             "date_joined": ["gt", "lt"],
+            "organizations__name": ["exact"],
```
Comment (dankolbman) [inline creator/users/schema.py:200]: As you can see, graphene converts these to strangely formatted parameters like `organizations_Name`. I'd like to avoid these by explicitly defining the fields on the filter where you can control the name.

File: tests/users/test_queries.py (hunk 1 above)
Comment (dankolbman) [reply tests/users/test_queries.py:13]: This should be supported by simply plugging in `organization` into the filterset. If not, you can use `GlobalIDMultipleChoiceFilter` to create the filter field.

**Review by dankolbman — APPROVED** [review]
# 🛺

File: tests/users/test_queries.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+from django.contrib.auth import get_user_model
+from creator.organizations.factories import OrganizationFactory
+from creator.organizations.models import Organization
+from creator.users.factories import UserFactory
+
+User = get_user_model()
+
+ALL_USERS_ORGANIZATION = """
+query ($organizations: [ID]) {
```
Comment (shanewilson) [inline tests/users/test_queries.py:13]: can we make this type `$organizations: [ID!]`? If you add a value to the list we shouldn't allowed it to be `null`

File: tests/users/test_queries.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  allUsers (organizations: $organizations){
```
Comment (shanewilson) [inline tests/users/test_queries.py:14]: just a note - allStudies only allows a single organization id. Allowing a list is probably more flexible, but we should make sure to align the other endpoints to match.

Comment (gsantia) [reply tests/users/test_queries.py:14]: honestly, i wasn't able to figure out how to do it in the case of a single org id, because it's a many-to-many relationship. i agree, i can add that functionality to `allStudies`

Comment (dankolbman) [reply tests/users/test_queries.py:14]: I'd agree with being consistent with other queries. Allowing to filter by multiple orgs would be nice, but we don't have that need right now and it's not how we do things elsewhere. You can probably achieve this pretty easily by swapping out to `GlobalIDFilter` instead of the multiple choice.

File: creator/users/schema.py [hunk 6]
```diff
@@ -180,6 +192,8 @@ def get_node(cls, info, id):
 
 
 class UserFilter(django_filters.FilterSet):
+    organizations = GlobalIDFilter(field_name="organizations")
```
Comment (shanewilson) [inline creator/users/schema.py:195]: i think these should be `organization` now right?

```suggestion
    organization = GlobalIDFilter(field_name="organization")
```

Comment (gsantia) [reply creator/users/schema.py:195]: true

**Review by shanewilson — APPROVED** [review]

File: tests/users/test_queries.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+def test_query_all_users_organization(
+    db, permission_client, permissions, allowed
+):
+    """
+    Test the allUsers query for filtering by organization name functionality.
+    """
+    user, client = permission_client(permissions)
+
+    # Setup users and organizations
+    org1 = OrganizationFactory()
```
Comment (znatty22) [inline tests/users/test_queries.py:49]:
```suggestion
    org1, org2, org3 = OrganizationFactory.create_batch(3)
```

File: tests/users/test_queries.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    org2 = OrganizationFactory()
+    org3 = OrganizationFactory()
+
+    user1 = UserFactory()
```
Comment (znatty22) [inline tests/users/test_queries.py:53]:
```suggestion
    user1, user2, user3 = UserFactory.create_batch(3)
```

File: tests/users/test_queries.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    user2 = UserFactory()
+    user3 = UserFactory()
+    for _ in range(10):
```
Comment (znatty22) [inline tests/users/test_queries.py:56]:
```suggestion
    UserFactory.create_batch(10)
```

**Review by znatty22 — CHANGES_REQUESTED** [review]

**Comment by gsantia** [comment]
Update on this, I've been stuck with circular imports lately and cannot figure out how to solve them without refactoring organizations and users. In another PR I added a `users` feature to `organization` queries, letting you see all the users in an organization. This required adding `UserNode` and `UserFilter` to the `OrganizationNode` definition.

But then in this PR we want to basically do the opposite, and we need `OrganizationNode` and `OrganizationFilter` to show up in the definition of `UserNode`. This leads to circular imports. At this point I've tried every trick I can think of and also whatever suggestions I can find online. Delayed imports don't work because they're both class definitions which execute immediately. Am I missing something? Otherwise I can do a bit of refactoring here and present it.

Basically I'd need to define `OrganizationNode` and `UserNode` in the same file and make all the changes that would entail. Not sure where this file would live. Probably in the organization app? Does this make sense?

**Comment by shanewilson** [comment]
> Update on this, I've been stuck with circular imports lately and cannot figure out how to solve them without refactoring organizations and users. In another PR I added a `users` feature to `organization` queries, letting you see all the users in an organization. This required adding `UserNode` and `UserFilter` to the `OrganizationNode` definition.
> 
> But then in this PR we want to basically do the opposite, and we need `OrganizationNode` and `OrganizationFilter` to show up in the definition of `UserNode`. This leads to circular imports. At this point I've tried every trick I can think of and also whatever suggestions I can find online. Delayed imports don't work because they're both class definitions which execute immediately. The Django importing with strings pattern also doesn't work because these are not Model classes. Am I missing something? Otherwise I can do a bit of refactoring here and present it.
> 
> Basically I'd need to define `OrganizationNode` and `UserNode` in the same file and make all the changes that would entail. Not sure where this file would live. Probably in the organization app? Does this make sense?

hey @gsantia you should be able to solve this by either using a "string" or a lambda. For example here https://github.com/kids-first/kf-api-study-creator/pull/703/files#diff-c4b79ed7497d7b73defcc01c8ca2a43aed837b3cd4a5c678e923ceffbdc00401R103 if you change it to the string `"OrganizationNode"` graphene should lazy load the model. If that doesn't work you may need to use the full path like `"creator.organizations.nodes.OrganizationNode"`.

https://stackoverflow.com/questions/62773292/graphql-design-with-circular-dependency
https://github.com/graphql-python/graphene/blob/a53b782bf8ec5612d5cceb582fbde68eeba859aa/graphene/utils/module_loading.py#L5

**Comment by gsantia** [comment]
> hey @gsantia you should be able to solve this by either using a "string" or a lambda. For example here https://github.com/kids-first/kf-api-study-creator/pull/703/files#diff-c4b79ed7497d7b73defcc01c8ca2a43aed837b3cd4a5c678e923ceffbdc00401R103 if you change it to the string `"OrganizationNode"` graphene should lazy load the model. If that doesn't work you may need to use the full path like `"creator.organizations.nodes.OrganizationNode"`.
> 
> https://stackoverflow.com/questions/62773292/graphql-design-with-circular-dependency
> https://github.com/graphql-python/graphene/blob/a53b782bf8ec5612d5cceb582fbde68eeba859aa/graphene/utils/module_loading.py#L5
> 
> if none of that works try `lambda: OrganizationNode` or `lambda: creator.organizations.nodes.OrganizationNode`

thanks for the suggestions! unfortunately i've tried these ideas and several variations of them and it still won't work. i think the key issue is that organization needs to use `UserFilter` now and user needs to use `OrganizationFilter`. so they're more tightly-coupled than just between node definitions. i'm trying to put them all into one file right now.

**Comment by shanewilson** [comment]
Are you able to use a "string" or a lambda on the Filters as well to decouple them? If we can't this seems like a major regression in Graphene since I know we used to be able to handle this in an earlier version.

**Comment by gsantia** [comment]
> Are you able to use a "string" or a lambda on the Filters as well to decouple them? If we can't this seems like a major regression in Graphene since I know we used to be able to handle this in an earlier version.

maybe I'm doing it wrong, but it doesn't seem to be working. i think i found a solution though and will push it soon

File: creator/organizations/mutations.py [hunk 10]
```diff
@@ -5,14 +5,14 @@
...
-from creator.organizations.nodes import OrganizationNode
+from creator.organizations.queries import OrganizationNode
 from creator.events.models import Event
 
 User = get_user_model()
 
 
 class CreateOrganizationInput(graphene.InputObjectType):
-    """ Schema for creating a new study """
+    """Schema for creating a new study"""
```
[the remark this answers is not in the document]
Comment (gsantia) [reply creator/organizations/mutations.py:15]: yeah sorry it was black

**Review by dankolbman — APPROVED** [review]
# 🍙

**Review by znatty22 — APPROVED** [review]
👍
